pkg/scheduler/controller/shoot/scheduler_control.go

```diff
@@ -218,14 +228,35 @@ func filterSeedsMatchingSeedSelector(cloudProfile *gardencorev1beta1.CloudProfil
 	return matchingSeeds, nil
 }
 
-func getCandidates(shoot *gardencorev1beta1.Shoot, seedList []*gardencorev1beta1.Seed, strategy config.CandidateDeterminationStrategy) ([]*gardencorev1beta1.Seed, error) {
+func filterSeedsMatchingProviders(cloudProfile *gardencorev1beta1.CloudProfile, shoot *gardencorev1beta1.Shoot, seedList []*gardencorev1beta1.Seed) ([]*gardencorev1beta1.Seed, error) {
+	var providers []string
+	if cloudProfile.Spec.SeedSelector != nil {
+		providers = cloudProfile.Spec.SeedSelector.Providers
+	}
+
+	var matchingSeeds []*gardencorev1beta1.Seed
+	for _, seed := range seedList {
+		if seed.DeletionTimestamp == nil && matchProvider(seed.Spec.Provider.Type, shoot.Spec.Provider.Type, providers) && !gardencorev1beta1helper.TaintsHave(seed.Spec.Taints, gardencorev1beta1.SeedTaintInvisible) && common.VerifySeedReadiness(seed) {
+			matchingSeeds = append(matchingSeeds, seed)
+		}
+	}
+	return matchingSeeds, nil
+}
+
+func getCandidates(cloudProfile *gardencorev1beta1.CloudProfile, shoot *gardencorev1beta1.Shoot, seedList []*gardencorev1beta1.Seed, strategy config.CandidateDeterminationStrategy) ([]*gardencorev1beta1.Seed, error) {
 	var candidates []*gardencorev1beta1.Seed
 
 	switch {
 	case shoot.Spec.Purpose != nil && *shoot.Spec.Purpose == gardencorev1beta1.ShootPurposeTesting:
 		candidates = determineCandidatesOfSameProvider(seedList, shoot)
 	case strategy == config.SameRegion:
 		candidates = determineCandidatesWithSameRegionStrategy(seedList, shoot)
+	case strategy == config.BestRegion:
+		candidates = determineCandidatesWithSameRegionStrategy(seedList, shoot)
+		if len(candidates) > 0 {
+			return candidates, nil
+		}
```

Comment:
```suggestion
				filtered := determineCandidatesWithSameRegionStrategy(seedList, shoot)
		if len(filtered) > 0 {
			return candidates, nil
		}
```
This is to not override the candidates so that you can have the provider check in the determineCandidatesWithSameRegionStrategy().
Also see my other comment there.

Reply:
It does not overwrite anything. Every *case* set the `candidates` variable. It is not filtered or reused, therefore there is no reason to use another variable for a dedicated case.